### PR TITLE
Implement Session window

### PR DESCRIPTION
**Pull request by garcher22**
Implement a new type of frame called SESSION for window functions that includes consecutive events separated by no longer than a given timeout. The syntax is like `window w as (order by time session 60)`, and this means a session window with timeout of 60 seconds.

This is an approach to implementing some parts of https://github.com/ClickHouse/ClickHouse/issues/45311

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Implement Session Window for window functions.

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit fd952b70bd7df70e41fb45baeb2d77cc1850507c with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/56099/fd952b70bd7df70e41fb45baeb2d77cc1850507c/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>❌ failure</td></tr>
<tbody>
</table>

**Review by vdimir — COMMENTED** [review]
Thank you for your PR!

I skimmed trough, couple of questions.
Also not sure about naming. Can we check out similar functionality in other systems?

File: src/Client/QueryFuzzer.cpp [hunk 1]
```diff
@@ -394,6 +394,12 @@ void QueryFuzzer::fuzzWindowFrame(ASTWindowDefinition & def)
             def.frame_type = r == 0 ? WindowFrame::FrameType::ROWS
                 : r == 1 ? WindowFrame::FrameType::RANGE
                     : WindowFrame::FrameType::GROUPS;
+            if (def.frame_type == WindowFrame::FrameType::SESSION)
+            {
+                // Have to initialize the window threshold when switching to
+                // SESSION frame, so that the formatting doesn't segfault.
```
[the remark this answers is not in the document]
Comment (vdimir) [reply src/Client/QueryFuzzer.cpp:400]: How we can get this AST? Shouldn't parser reject illegal input then?

File: src/Processors/Transforms/WindowTransform.cpp [hunk 2]
```diff
@@ -325,6 +328,25 @@ WindowTransform::WindowTransform(const Block & input_header_,
...
+        {
+            window_description.frame.session_window_threshold = convertFieldToTypeOrThrow(
+                window_description.frame.session_window_threshold,
+                *entry.type);
+
+            /*
+             * Note that the value can be NaN, and we don't have a visitor
+             * for "greater", so we have to write the condition in double negation style.
+             */
+            if (!applyVisitor(FieldVisitorAccurateLess{}, Field(0),
```
[the remark this answers is not in the document]
Comment (vdimir) [reply src/Processors/Transforms/WindowTransform.cpp:342]: For example for dates, I don't think some DateTime make sense, it should be rather Interval data type.

**Comment by UnamedRus** [comment]
> But I think ClickHouse doesn't have any notion of grouping over ranges of ordered data in its GROUP BY executor, and implementing it would be a major effort.

There was an idea to implement GROUP BY for windows/sessions

https://github.com/ClickHouse/ClickHouse/issues/42194 "Fuzzy GROUP BY for data clustering"

**Comment by UnamedRus** [comment]
Ordered set aggregations is another beast IMO, and i still not completely sure about them.

How many order depended aggregate functions do we have actually?
groupArray family, 
some otherы which just somewhat unstable because of order (but we don't count/care about them)
What else?

**Comment by UnamedRus** [comment]
> area under curve divided by total time

BTW, there is arrayAUC function, may be it can be used for that somehow.

**Comment by UnamedRus** [comment]
BTW, we also will need special window analytical function to return some `session_id`, which can be used later in group by to collect all rows related to single session..

**Comment by UnamedRus** [comment]
> This can be done with min(<order by column>) over (...). It gives you the start position

This make sense, it can work as W/A for time being.

**Comment by qoega** [comment]
@groeneai can you check if we can update this PR to latest master? Can you review implementation and check if implementation does not contradict with current approaches?

**Comment by groeneai** [comment]
@qoega I went through the PR against current master (`59d2926`). Summary below.

## 1. Can it be updated to latest master? Yes, easily.

The branch is ~85k commits behind (base is Dec 2024), but a trial merge of `origin/master` produces only **4 conflicts, all trivial**; the other 10 touched files auto-merge cleanly. The `WindowFrame` struct itself is essentially unchanged since the PR's base, so the additions still apply.

| Conflicting file | Nature | Resolution |
|---|---|---|
| `docs/.../window-functions/index.md` | doc drift | keep both, rewrite the section (see 2.1) |
| `src/Analyzer/WindowNode.cpp` | master added a `hasAlias()` branch to `toASTImpl`; PR added a SESSION branch | keep both blocks |
| `src/Common/QueryFuzzer.cpp` | master reformatted `case 0`; PR replaced it with a `frame_types` array incl. SESSION | take PR side |
| `src/Processors/Transforms/WindowTransform.cpp` | master added a `should_materialize` block + turned `assert`->`chassert`; PR added `SESSION ||` to the same two conditions | mechanical, apply PR's `SESSION` to master's versions |

So updating is a short mechanical rebase, no structural rework required.

## 2. Does the implementation contradict current approaches?

The core algorithm is sound and reuses the existing RANGE-OFFSET machinery correctly: single ORDER BY key enforced in `checkValid`, `compare_values_with_offset` with `offset_is_preceding = direction < 0` for DESC, streaming/partition-end handling in `advanceFrameEndSession` mirrors `advanceFrameEndRangeOffset`, and `arePeers` extended so equal-key rows share a frame. Threshold units follow RANGE-OFFSET semantics (natural unit of the ORDER BY column). That part is consistent.

The open questions are less about the code and more about the design, which is the same thing vdimir raised in 2023 and was never settled:

**2.0 (design, needs a maintainer call) SESSION is a disjoint-frame construct, not a sliding/growing frame.** As garcher22 noted himself, it is "logically closer to grouping than windowing": every row in a session shares one frame, and `begin_type`/`begin_offset`/`end_type`/`end_offset` are all meaningless for it yet still carried in `WindowFrame`. Overloading `FrameType` works, but before merging it is worth deciding whether this is the intended abstraction versus (a) a `session_id`-producing analytic function usable in `GROUP BY` (what UnamedRus asked for), or (b) folding into the "Fuzzy GROUP BY" idea (#42194). This is the main "does it contradict current approaches" point and it is a maintainer decision.

Concrete, mergeable-blocking items:

**2.1 Docs are a placeholder.** `index.md` literally contains "add some proper docs". Needs a real section: syntax, that the threshold is in the ORDER BY column's units (so "60-second window" only holds for a DateTime/numeric-seconds key), DESC behaviour, and an example.

**2.2 `WindowFrame::toString()` is not patched for SESSION.** The PR patches the AST formatter (`ASTWindowDefinition::formatImpl`) but not the runtime `WindowFrame::toString(WriteBuffer&)` in `WindowDescription.cpp`. That function unconditionally emits `type << " BETWEEN " ... " AND " ...`, so a SESSION frame renders as the misleading `SESSION BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW` in `dump()` / error messages. (`buf << type` itself is fine, magic_enum prints `SESSION`.) It should print the threshold instead.

**2.3 No experimental gate.** This is a new, non-standard SQL surface and a new execution strategy. ClickHouse gates such surfaces behind `allow_experimental_*` (e.g. `allow_experimental_window_view`). A setting like `allow_experimental_session_window_frame`, defaulting off, would let it roll out safely and be rejected by default until stabilized.

Smaller items:
- **INTERVAL support / units:** the parser uses `ParserExpression`, so `SESSION INTERVAL 5 MINUTE` should parse, but the test only covers plain numerics. Please add coverage (or explicitly reject INTERVAL) so the DateTime-key case is pinned.
- **Test is thorough on error paths** (serverError 69/36/48/BAD_ARGUMENTS) and DESC/fractional cases, good. Consider adding a DateTime ORDER BY case and a PARTITION BY case, since those are the real-world uses.

## Bottom line
Rebase is easy (4 trivial conflicts). The executor logic is correct and consistent with the RANGE-OFFSET code. Before it can go in it needs: real docs, the `toString` fix, an experimental setting, and a decision from the maintainers on the disjoint-frame-as-`FrameType` design that was left open in the original 2023 review.

I did not push anything to the branch (external PR). Happy to expand on any of the above.

**Comment by garcher22** [comment]
@groeneai please do it for me, I have no time for this work anymore. You have permission to push to my branch.

**Comment by groeneai** [comment]
Thanks @garcher22, happy to carry this forward. Plan:

I'll open a continuation PR from a `groeneai` fork branch that rebases the session-window work onto current master and credits you as the original author (`Co-authored-by`), rather than force-pushing over your branch. That keeps your history intact and avoids the branch going stale again. It will reference and close this PR.

Scope for the continuation PR (from my 2026-07-01 review):
1. Mechanical rebase onto master (the 4 trivial conflicts noted earlier).
2. Fix `WindowFrame::toString(WriteBuffer&)` so a SESSION frame doesn't render as `SESSION BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW`.
3. Real docs section (syntax, threshold-in-ORDER-BY-units, DESC, example).
4. Gate behind `allow_experimental_session_window_frame` (default off).
5. Extend tests: DateTime/INTERVAL ORDER BY key + PARTITION BY.

One design point still needs a maintainer call and I'll flag it prominently on the new PR: SESSION is a disjoint-frame construct (every row in a session shares one frame), which is closer to grouping than windowing. Overloading `FrameType` works and the executor reuses the RANGE-OFFSET machinery correctly, but whether this is the intended abstraction (vs a `session_id`-producing analytic function usable in `GROUP BY`) was left open in the 2023 review. I'll get it in front of the window-functions maintainers on the rebased, CI-green PR so the discussion happens on live code.

I'll link the new PR here once it's up.

**Comment by groeneai** [comment]
Opened the continuation PR: #109672 (from `groeneai:session-window-continuation`, rebased onto master, your commits credited via Co-authored-by). It resolves the merge conflicts, adapts to the current AST/formatter/accurateLess APIs, gates the feature behind `allow_experimental_session_window_frame`, fixes `WindowFrame::toString` for the SESSION frame, adds docs, and extends the test. Review is running there now.
